core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -261,14 +283,54 @@ final class ShuffleBlockFetcherIterator(
             results.put(new SuccessFetchResult(BlockId(blockId), infoMap(blockId)._2,
               address, infoMap(blockId)._1, buf, remainingBlocks.isEmpty))
             logDebug("remainingBlocks: " + remainingBlocks)
+            enqueueDeferredFetchRequestIfNecessary()
           }
         }
         logTrace(s"Got remote block $blockId after ${Utils.getUsedTimeNs(startTimeNs)}")
       }
 
       override def onBlockFetchFailure(blockId: String, e: Throwable): Unit = {
         logError(s"Failed to get block(s) from ${req.address.host}:${req.address.port}", e)
-        results.put(new FailureFetchResult(BlockId(blockId), infoMap(blockId)._2, address, e))
+        val (size, mapIndex) = infoMap(blockId)
+        e match {
+          // SPARK-27991: Catch the Netty OOM and set the flag `isNettyOOMOnShuffle` (shared among
+          // tasks) to true as early as possible. Unless there's no in-flight requests, the pending
+          // fetch requests won't be raised afterwards until the flag is set to false on:
+          // 1) the Netty free memory >= average remote block size - we'll check this whenever
+          //    there's a fetch request succeeds.
+          // 2) the number of in-flight requests becomes 0 - we'll check this in `fetchUpToMaxBytes`
+          //    whenever it's invoked.
+          // Although Netty memory is shared across multiple modules, e.g., shuffle, rpc, the flag
+          // only takes effect for the shuffle due to the implementation simplicity concern.
+          // And we'll buffer the consecutive block failures caused by the OOM error until there's
+          // no remaining blocks in the current request. Then, we'll package these blocks into
+          // a same fetch request for the retry later. In this way, instead of creating the fetch
+          // request per block, it would help reduce the concurrent connections and data loads
+          // pressure at remote server.
+          // Note that catching OOM and do something based on it is only a workaround for
+          // handling the Netty OOM issue, which is not the best way towards memory management.
+          // We can get rid of it when we find a way to manage Netty's memory precisely.
+
+          // Ensure the Netty memory is at least enough for serving only one block to avoid
+          // the endless retry. And since the Netty memory is shared among multiple modules,
+          // we use the factor "1.5" for the overhead concern.
+          case _: OutOfDirectMemoryError if PlatformDependent.maxDirectMemory() > ( 1.5 * size) =>
```

Comment:
I wonder using the `size` maybe not accurate for the block that bigger than `maxReqSizeShuffleToMem`. For example, Netty memory is 4G, and the block is 4G too. But it doesn't necessarily mean the block needs 4G netty memory as it would be transferred to the disk during fetching. 

@mridulm @otterc @attilapiros @cloud-fan  do you have any ideas about this?

Reply:
What if we simply remove this check? Will we hang?

Reply:
I do not think the condition is needed.

My reasoning:
1. PlatformDependent.maxDirectMemory() is a constant value: either set as `-Dio.netty.maxDirectMemory <value>` or get from the JVM: 
https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L160
2. Usually we have the 2nd case.
3. maxDirectMemory is usually way a higher then `maxReqSizeShuffleToMem`. At least the defaults: maxReqSizeShuffleToMem=200M, [maxDirectMemory](https://dzone.com/articles/default-hotspot-maximum-direct-memory-size)
 
But another idea to introduce a timestamp how long we are deferring fetch requests. That way if we hang that won't be forever.

Reply:
> What if we simply remove this check? Will we hang?

Theoretically, it will hang when the netty max memory is not even to serve a single block.

Reply:
@attilapiros  I agree that we won't hang normally given the truth that Netty memory is usually larger than the block.

Limit time or the number of retries both should works.

Reply:
As I see to eliminate the unusual cases we should not focus on the block sizes and on the defer feature only but we should consider validating "spark.network.maxRemoteBlockSizeFetchToMem" against maxDirectMemory when "spark.shuffle.io.preferDirectBufs" is on. 

To handle the case (without the defer feature) when preferDirectBufs is true and maxDirectMemory is less than maxRemoteBlockSizeFetchToMem and a block comes which bigger than maxDirectMemory but less than maxRemoteBlockSizeFetchToMem.

WDYT?

Reply:
Make sense. I think we should fail fast in this case during the initialization of `ShuffleBlockFetcherIterator`.